crates/driver/src/domain/competition/sorting.rs

```diff
@@ -0,0 +1,122 @@
+use {
+    crate::{
+        domain::{
+            competition::{auction::Tokens, order},
+            eth,
+        },
+        util,
+    },
+    chrono::{Duration, Utc},
+    std::{cmp::Ordering, sync::Arc},
+};
+
+pub trait OrderComparator: Send + Sync {
+    fn compare(
+        &self,
+        order_a: &order::Order,
+        order_b: &order::Order,
+        tokens: &Tokens,
+        solver: &eth::H160,
+    ) -> Ordering;
+}
+
+impl<F> OrderComparator for F
+where
+    F: Fn(&order::Order, &order::Order, &Tokens, &eth::H160) -> Ordering + Send + Sync,
+{
+    fn compare(
+        &self,
+        order_a: &order::Order,
+        order_b: &order::Order,
+        tokens: &Tokens,
+        solver: &eth::H160,
+    ) -> Ordering {
+        self(order_a, order_b, tokens, solver)
+    }
+}
+
+pub trait OrderingKey: Send + Sync + 'static {
+    type Key: Ord + Send + Sync + 'static;
+
+    fn key(&self, order: &order::Order, tokens: &Tokens, solver: &eth::H160) -> Self::Key;
+
+    /// Returns a comparator that compares two orders based on the key in
+    /// reverse order.
+    fn into_comparator(self) -> Arc<dyn OrderComparator>
+    where
+        Self: Sized,
+    {
+        Arc::new(
+            move |a: &order::Order, b: &order::Order, tokens: &Tokens, solver: &eth::H160| {
+                self.key(a, tokens, solver)
+                    .cmp(&self.key(b, tokens, solver))
+                    .reverse()
```

Comment:
Previously the `likelihood` was computed once and cached. This sped up sorting from `1.5s` to `0.2` when it was implemented. Since we now have many more sorting options with `likelihood` probably being one of the last ones it might not be a problem but it would be good to measure before merging.
The easiest way to do it would be to run the shadow environment locally.

Reply:
This is not possible to run shadow competition locally. More details: [[1](https://github.com/cowprotocol/services/pull/2855#discussion_r1705376614), [2](https://github.com/cowprotocol/services/pull/2855#discussion_r1707242997)]

Reply:
Will try to create a cache inside the sorting function manually.

Reply:
You should be able to use `sort_by_cached_key()`. For that you would compute all the sorting keys for each of the sorting strategies once, shove them into a vector and implement `Ord` on that by comparing each key one after the other.
I would assume that's the easiest way to make use of `sort_by_cached_key()` but would be suboptimal because there might be cases where the expensive `likelihood` function would never have to be called for some orders because the earlier predicates already suffice to determine the order.

Reply:
> I would assume that's the easiest way to make use of sort_by_cached_key() but would be suboptimal because there might be cases where the expensive likelihood function would never have to be called for some orders because the earlier predicates already suffice to determine the order.

I have already considered a similar approach but didn't know about the `1.5s` sorting issue. `sort_by_cached_key` should provide less implementation and time complexity. https://github.com/cowprotocol/services/pull/2872/commits/d6e6878f12e4b43f2121408e73631fdcbab10e94